### DISCUSSION TITLE
Vehicle Routing Problem - Dynamic Time Windows

### DISCUSSION THREAD

**Discussion by edders51** [google/or-tools#3163] [Routing (and legacy CP) questions]
Hi,

I am new to or-tools and trying to solve the following problem (somewhat idealised):

A number of maintenance jobs of varying lengths in varying locations must be performed. One or more buses will drop off and pick up the workers (1 per job) throughout the day, shuttling them around to get all the jobs done. The amount of time they need to spend on each job is known.

I think I can approach this as a VRPTW if I duplicate nodes to represent the drop off and pick up of a worker - A1, A2, B1, B2 etc. 

The problem I have is I cannot set the time window for pick up (note that pick up is after drop off in my problem conceptually) until I know what time the worker was dropped off.

I presume there is a solution using constraints but i'm struggling to find it. Perhaps using time windows is not the way here?

Any help or pointers gratefully received.

Cheers,
Ed

**Comment by lperron** [comment]
Could also be seen as a flexible scheduling problem with setup times.

**Reply by edders51** [reply]
Thank you for the suggestion.

I'm not sure if it will work though.
Taking the example from scheduling_with_transitions_sat.py if i treat the machine as the worker, so:
- number of machines = number of workers 
- machine time = job time
I can treat the transition time as the journey from one location to the next, which i can implement with a matrix.

The problem here is that N_buses < N_workers, so whilst i can check whether the workers can complete the designated tasks, i cannot determine how many buses i need.

Is there any way to add a further layer to the scheduling problem to represent the number of buses constraint? I have not been able to find any similar examples.
I presume that the way the VRPTW solver works is it takes the time windows as a hard input and they cannot be adjusted on the fly?

Many thanks!

**Reply by lperron** [reply]
for each arc of the transition matrix, you know the start and end (end time of the previous job, start time of the next job), and the duration.
So you could create one optional interval var for each transition, make it fit in the time windows, and fix the duration of the transit.
Then add all optional arcs in a cumulative constraint of capacity num buses.

Now, nothing makes sure the assignment of buses is feasible. So, you need to create a graph containing all transitions, then do a vrp constraint (MultipleCircuitConstraint) on it. not entirely satisfactory. 

Another issue is to model workers sharing the same bus.

And no, time windows in the routing library are static. You could duplicate nodes with different time windows, but not sure it helps.

**Reply by edders51** [reply]
Having spent a bit longer getting used to the tool, i have decided that you were correct when you suggested flexible scheduling problem with setup times, so i am working on that avenue - thanks.
